torch/utils/_python_dispatch.py

```diff
@@ -51,6 +53,20 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         _pop_mode()
 
+    def check_mode_state_push(self):
+        if not hasattr(self, "tracking"):
+            self.tracking = threading.local()
+        else:
+            if hasattr(self.tracking, "on_stack"):
+                assert self.tracking.on_stack is False, "Illegal attempt to push an already pushed mode onto the stack"
+
+        self.tracking.on_stack = True
+
+    def check_mode_state_pop(self):
+        assert hasattr(self, "tracking"), "Unexpected, popping a mode we are not tracking"
+        assert self.tracking.on_stack is True, "Unexpected, popping a mode that thinks its already off the stack"
+        self.tracking.on_stack = False
```

Comment:
Per @voznesenskym's doc + question:
This is where we're adding the information per-instance of the class

In terms of big picture what we want to do, happy to hash this out before or after landing (depending on when I can fix this segfault). Making this exist per class would just involve making tracking exist on the class and turn these into class methods